### PR TITLE
simplify generate

### PR DESCRIPTION
**Pull request by johnynek**
this refactors most of #332 and #89 so that check-only works for anything using the IO abstraction.

This allows check to just work for target files, resolution jsons, and poms.

cc @kevingessner

### CODE REVIEW

**Comment by johnynek** [comment]
note the CI did catch that [e92579e](https://github.com/bazeltools/bazel-deps/pull/335/commits/e92579ebd5649e19c7b471fe95762a7ef4f9ab19) was broken.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -22,4 +22,5 @@ jobs:
       - name: run tests
         run: |
           ./bazel build //...
-          ./bazel test --test_output=errors //... --test_env=COURSIER_CACHE="${{ github.workspace }}/.coursier-cache"
\ No newline at end of file
+          ./bazel test --test_output=errors //... --test_env=COURSIER_CACHE="${{ github.workspace }}/.coursier-cache"
+          ./bazel run //:parse -- generate -r `pwd` -s 3rdparty/workspace.bzl -d dependencies.yaml -t 3rdparty/target_file.bzl -p pom.xml --check-only
```
Comment (eed3si9n) [inline .github/workflows/ci.yml:26]: Is this `--check-only` feature documented in README?

File: src/scala/com/github/johnynek/bazel_deps/IO.scala [hunk 2]
```diff
@@ -115,83 +111,194 @@ object IO {
...
     }
 
+  def orConst[A](optIO: Option[Result[A]], a: A): Result[A] =
+    optIO match {
+      case Some(io) => io
+      case None => IO.const(a)
+    }
+
+  def orUnit(optIO: Option[Result[Unit]]): Result[Unit] =
+    orConst(optIO, ())
```
Comment (wrobstory) [inline src/scala/com/github/johnynek/bazel_deps/IO.scala:121]: Is `orConst` used anywhere else? Wondering if we even need the separate method if we're only ever using `orUnit`

**Review by eed3si9n — APPROVED** [review]

File: src/scala/com/github/johnynek/bazel_deps/IO.scala [hunk 3]
```diff
@@ -115,83 +111,194 @@ object IO {
...
+      case ReadFile(f) =>
+        Try({
+          val ff = fileFor(f)
+          if (ff.exists)
+            Some(new String(Files.readAllBytes(ff.toPath), charset))
+          else None
+        })
+      case Failed(err) => Failure(err)
+      // The following are not read only
+      case MkDirs(f) => Try(!fileFor(f).exists())
```
Comment (wrobstory) [inline src/scala/com/github/johnynek/bazel_deps/IO.scala:145]: So in a read only case this is kinda overloading the `MkDirs` op to check for directory existence?

File: src/scala/com/github/johnynek/bazel_deps/IO.scala [hunk 4]
```diff
@@ -115,83 +111,194 @@ object IO {
...
+  }
+
+  class ReadWriteExec(root: File) extends ReadOnlyExec(root) {
+    override def apply[A](o: Ops[A]): Try[A] = o match {
+      case MkDirs(f) => Try(fileFor(f).mkdirs())
+      case RmRf(f, removeHidden) =>
+        Try {
+          // get the java path
+          val file = fileFor(f)
+          // require(file.isDirectory, s"$f is not a directory")
```
Comment (wrobstory) [inline src/scala/com/github/johnynek/bazel_deps/IO.scala:159]: Can we remove this comment

File: src/scala/com/github/johnynek/bazel_deps/MakeDeps.scala [hunk 5]
```diff
@@ -88,16 +88,61 @@ object MakeDeps {
...
+          g.shaFilePath.map(IO.path(_)),
+          g.targetFile.map(IO.path(_)),
+          g.enable3rdPartyInRepo,
+          ws,
+          targets,
+          formatter,
+          g.resolvedOutput.map(IO.path),
+          artifacts) >> pomIO
+
+        if (g.checkOnly) {
```
Comment (eed3si9n) [inline src/scala/com/github/johnynek/bazel_deps/MakeDeps.scala:104]: This is a pretty long `if {...}`. Maybe pull the content into a new function that returns an IO?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (johnynek) [reply .github/workflows/ci.yml:26]: it is not currently. Only the cli help it looks like.

File: src/scala/com/github/johnynek/bazel_deps/MakeDeps.scala [hunk 6]
```diff
@@ -15,7 +15,7 @@ object MakeDeps {
 
   private[this] val logger = LoggerFactory.getLogger("MakeDeps")
 
-  def apply(g: Command.GenerateLike): Unit = {
+  def apply(g: Command.Generate): Unit = {
```
Comment (eed3si9n) [inline src/scala/com/github/johnynek/bazel_deps/MakeDeps.scala:18]: Wrap the whole thing in IO?

File: src/scala/com/github/johnynek/bazel_deps/IO.scala [hunk 7]
```diff
@@ -115,83 +111,194 @@ object IO {
...
+  }
+
+  sealed abstract class CheckException(val message: String) extends Exception(message) {
+    def path: IO.Path
+  }
+
+  object CheckException {
+    case class DirectoryMissing(path: IO.Path) extends CheckException(s"directory ${path.asString} expected but missing")
+    case class WriteMismatch(path: IO.Path, current: Option[String], expected: String, compressed: Boolean) extends CheckException(
+      (if (compressed) "compressed " else "") + s"file ${path.asString} didn't match."
```
Comment (wrobstory) [inline src/scala/com/github/johnynek/bazel_deps/IO.scala:225]: You don't do anything with `current` and `expected` here, but I think you should! 😄 Particularly on the case where `current` is None, I think you should probably have the error return something like `file {path} does not exist!`

File: src/scala/com/github/johnynek/bazel_deps/MakeDeps.scala [hunk 8]
```diff
@@ -88,16 +88,61 @@ object MakeDeps {
...
+        val io = executeGenerate(
+          model,
+          g.shaFilePath.map(IO.path(_)),
+          g.targetFile.map(IO.path(_)),
+          g.enable3rdPartyInRepo,
+          ws,
+          targets,
+          formatter,
+          g.resolvedOutput.map(IO.path),
+          artifacts) >> pomIO
```
Comment (wrobstory) [inline src/scala/com/github/johnynek/bazel_deps/MakeDeps.scala:102]: TIL about the `IO` `>>` operator

**Review by wrobstory — CHANGES_REQUESTED** [review]

File: src/scala/com/github/johnynek/bazel_deps/IO.scala (hunk 3 above)
Comment (johnynek) [reply src/scala/com/github/johnynek/bazel_deps/IO.scala:145]: mkdirs has to return if it made the directory or not. We aren't making it, but we *would* make it in a phantom write situation if the directory is absent.

That said, I don't think we need to get cute. I think we can just fail all write operations in read-only mode.

File: src/scala/com/github/johnynek/bazel_deps/MakeDeps.scala (hunk 6 above)
Comment (johnynek) [reply src/scala/com/github/johnynek/bazel_deps/MakeDeps.scala:18]: I want to push this until I make a follow up PR that actually uses cats-effect.

**Review by wrobstory — APPROVED** [review]

**Comment by johnynek** [comment]
closes #333

**Pull request merged**